### PR TITLE
Add support for Razer BlackWidow V4

### LINKED ISSUE #2163: Support for Razer BlackWidow V4
### Device Info

* Name: BlackWidow v4
* Product Number: RZ03-04692100-R3G1
* Model Number: RZ03-0469

### Device Information
```
Please post the output of
    lsusb -d 1532: -v

`Bus 005 Device 007: ID 1532:0287 Razer USA, Ltd Razer BlackWidow V4
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0287 Razer BlackWidow V4
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V4
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  bNumConfigurations      0
Device Status:     0x0000
  (Bus Powered)
`
```

```
Please post the output of
    ls -lah /dev/input/by-id/

total 0
drwxr-xr-x 2 root root 300 14. Nov 16:48 .
drwxr-xr-x 4 root root 660 14. Nov 16:48 ..
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Burr-Brown_from_TI_USB_Audio_CODEC-event-if03 -> ../event8
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-©Microsoft_Corporation_Controller_0A5101D-event-joystick -> ../event5
lrwxrwxrwx 1 root root   6 14. Nov 16:48 usb-©Microsoft_Corporation_Controller_0A5101D-joystick -> ../js0
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-event-mouse -> ../event2
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-if01-event-kbd -> ../event12
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-if02-event-kbd -> ../event14
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Razer_Razer_Basilisk_V3-mouse -> ../mouse0
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-event-if01 -> ../event22
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-event-if04 -> ../event24
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-event-kbd -> ../event20
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-if01-event-kbd -> ../event21
lrwxrwxrwx 1 root root  10 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-if02-event-mouse -> ../event23
lrwxrwxrwx 1 root root   9 14. Nov 16:48 usb-Razer_Razer_BlackWidow_V4-if02-mouse -> ../mouse1

```

```
Please post the output of
    usbhid-dump -m 1532 -ed

005:005:003:DESCRIPTOR         1699978231.860639
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 95 5A B1 01 C0

005:005:002:DESCRIPTOR         1699978231.863627
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

005:005:001:DESCRIPTOR         1699978231.866627
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 0C 09 01 A1 01 85
 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81
 00 75 08 95 0D 81 01 C0 05 01 09 80 A1 01 85 03
 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95 05
 81 01 75 08 95 0E 81 01 C0 05 01 09 00 A1 01 85
 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95
 0F 81 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00
 26 FF 00 35 00 46 FF 00 75 08 95 0F 81 00 C0

005:005:000:DESCRIPTOR         1699978231.869627
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 0B 81 03
 05 0C 0A 38 02 15 81 25 7F 75 08 95 01 81 06 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 C0

005:007:004:DESCRIPTOR         1699978231.871632
 05 59 09 01 A1 01 85 01 09 02 A1 02 09 03 15 00
 27 FF FF 00 00 75 10 95 01 B1 03 09 04 09 05 09
 06 09 07 09 08 15 00 27 FF FF FF 7F 75 20 95 05
 B1 03 C0 85 02 09 20 A1 02 09 21 15 00 27 FF FF
 00 00 75 10 95 01 B1 02 C0 85 03 09 22 A1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 01 B1 02 09 23
 09 24 09 25 09 27 09 26 15 00 27 FF FF FF 7F 75
 20 95 05 B1 02 09 28 09 29 09 2A 09 2C 09 2D 15
 00 26 FF 00 75 08 95 05 B1 02 C0 85 04 09 50 A1
 02 09 03 09 55 15 00 25 08 75 08 95 02 B1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 08 B1 02 09 51
 09 52 09 53 09 51 09 52 09 53 09 51 09 52 09 53
 09 51 09 52 09 53 09 51 09 52 09 53 09 51 09 52
 09 53 09 51 09 52 09 53 09 51 09 52 09 53 15 00
 26 FF 00 75 08 95 18 B1 02 C0 85 05 09 60 A1 02
 09 55 15 00 25 08 75 08 95 01 B1 02 09 61 09 62
 15 00 27 FF FF 00 00 75 10 95 02 B1 02 09 51 09
 52 09 53 15 00 26 FF 00 75 08 95 03 B1 02 C0 85
 06 09 70 A1 02 09 71 15 00 25 01 75 08 95 01 B1
 02 C0 85 07 06 00 FF 15 00 26 FF 00 09 02 75 08
 95 3F B1 02 C0 05 0C 09 01 A1 01 85 08 19 D0 29
 D7 15 00 25 01 75 01 95 08 81 02 C0

005:007:003:DESCRIPTOR         1699978231.874630
 05 01 09 00 A1 01 06 00 FF 09 03 15 00 25 01 75
 08 95 5A B1 01 C0

005:007:002:DESCRIPTOR         1699978231.877628
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 C0

005:007:001:DESCRIPTOR         1699978231.880631
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 29 A0 15 00 25 01
 75 01 95 A0 81 02 05 08 19 01 29 03 15 00 25 01
 75 01 95 03 91 02 95 05 91 01 C0 05 0C 09 01 A1
 01 85 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75
 10 81 00 75 08 95 13 81 01 C0 05 01 09 80 A1 01
 85 03 19 81 29 83 15 00 25 01 75 01 95 03 81 02
 95 05 81 01 75 08 95 14 81 01 C0 05 01 09 00 A1
 01 85 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75
 08 95 15 81 00 C0 05 01 09 00 A1 01 85 05 09 03
 15 00 26 FF 00 35 00 46 FF 00 75 08 95 15 81 00
 C0

005:007:000:DESCRIPTOR         1699978231.883631
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on **Razer Synapse** in a Windows Environment>

### PR DESCRIPTION
**Pull request by xavierog**
This pull request combines the initial work made in #2163 (the relevant commit retains its author) with a few extra lines related to the `M6` key.

Fixes #2163

### CODE REVIEW

**Comment by xavierog** [comment]
FWIW, the changes brought by this pull request are available as DEB packages: see details in https://github.com/openrazer/openrazer/issues/2163#issuecomment-2018226849

File: README.md [hunk 1]
```diff
@@ -94,6 +94,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade 15 Advanced (Early 2022)          |  1532:028A  |
 | Razer Blade 17 (2022)                         |  1532:028B  |
 | Razer Blade 14 (2022)                         |  1532:028C  |
+| Razer BlackWidow V4                           |  1532:0287  |
```
Comment (z3ntu) [inline README.md]: Keep this list sorted by USB PID, so move above 028A

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -1687,6 +1696,11 @@ static ssize_t razer_attr_write_matrix_effect_wheel(struct device *dev, struct d
     struct razer_report response = {0};
 
     switch(device->usb_pid) {
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V4:
+        request = razer_chroma_extended_matrix_effect_wheel(VARSTORE, BACKLIGHT_LED, direction);
+        request.transaction_id.id = 0x1F;
+        break;
+
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Why are you adding the code again? Just re-use the case for V4_PRO? Same comment for the other parts below in the diff.

File: driver/razerkbd_driver.h [hunk 3]
```diff
@@ -91,6 +91,7 @@
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_EARLY_2022 0x028A
 #define USB_DEVICE_ID_RAZER_BLADE_17_2022 0x028B
 #define USB_DEVICE_ID_RAZER_BLADE_14_2022 0x028C
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V4 0x0287
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: keep the list sorted by value

File: install_files/udev/99-razer.rules [hunk 4]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|0029|002e|002f|0032|0034|0036|0037|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|0118|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|0269|026a|026b|026c|026d|026e|026f|0270|0271|0276|0279|027a|0282|028a|028b|028c|028d|0290|0292|0294|0295|0296|0298|029e|029f|02a0|02a2|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|0118|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|0269|026a|026b|026c|026d|026e|026f|0270|0271|0276|0279|027a|0282|028a|028b|028c|0287|028d|0290|0292|0294|0295|0296|0298|029e|029f|02a0|02a2|0a24", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: keep the list sorted

File: README.md (hunk 1 above)
Comment (xavierog) [reply README.md]: Ok, fixed.

File: driver/razerkbd_driver.h (hunk 3 above)
Comment (xavierog) [reply driver/razerkbd_driver.h]: Ok, fixed.

File: install_files/udev/99-razer.rules (hunk 4 above)
Comment (xavierog) [reply install_files/udev/99-razer.rules]: Ok, fixed.

File: driver/razerkbd_driver.c (hunk 2 above)
Comment (xavierog) [reply driver/razerkbd_driver.c]: Ok, fixed all V4/V4_PRO duplicates.

**Review by z3ntu — APPROVED** [review]
Looks reasonable, thanks for the patch!

**Pull request merged**